core/chains/evm/client/node_fsm.go

```diff
@@ -0,0 +1,218 @@
+package client
+
+import (
+	"context"
+	"fmt"
+)
+
+// NodeState represents the current state of the node
+// Node is a FSM
+type NodeState int
+
+func (n NodeState) String() string {
+	switch n {
+	case NodeStateUndialed:
+		return "Undialed"
+	case NodeStateDialed:
+		return "Dialed"
+	case NodeStateInvalidChainID:
+		return "InvalidChainID"
+	case NodeStateAlive:
+		return "Alive"
+	case NodeStateUnreachable:
+		return "Unreachable"
+	case NodeStateOutOfSync:
+		return "OutOfSync"
+	case NodeStateClosed:
+		return "Closed"
+	default:
+		return fmt.Sprintf("NodeState(%d)", n)
+	}
+}
+
+// GoString prints a prettier state
+func (n NodeState) GoString() string {
+	return fmt.Sprintf("NodeState%s(%d)", n.String(), n)
+}
+
+const (
+	// NodeStateUndialed is the first state of a virgin node
+	NodeStateUndialed = NodeState(iota)
+	// NodeStateDialed is after a node has successfully dialed but before it has verified the correct chain ID
+	NodeStateDialed
+	// NodeStateInvalidChainID is after chain ID verification failed
+	NodeStateInvalidChainID
+	// NodeStateAlive is a healthy node after chain ID verification succeeded
+	NodeStateAlive
+	// NodeStateUnreachable is a node that cannot be dialed or has disconnected
+	NodeStateUnreachable
+	// NodeStateOutOfSync is a node that is accepting connections but exceeded
+	// the failure threshold without sending any new heads. It will be
+	// disconnected, then put into a revive loop and re-awakened after redial
+	// if a new head arrives
+	NodeStateOutOfSync
+	// NodeStateClosed is after the connection has been closed and the node is at the end of its lifecycle
+	NodeStateClosed
+	// nodeStateLen tracks the number of states
+	nodeStateLen
+)
+
+// allNodeStates represents all possible states a node can be in
+var allNodeStates []NodeState
+
+func init() {
+	for s := NodeState(0); s < nodeStateLen; s++ {
+		allNodeStates = append(allNodeStates, s)
+	}
+
+}
+
+// FSM methods
+
+// State allows reading the current state of the node
+func (n *node) State() NodeState {
+	n.stateMu.RLock()
+	defer n.stateMu.RUnlock()
+	return n.state
+}
+
+// setState is only used by internal state management methods.
+// This is low-level; care should be taken by the caller to ensure the new state is a valid transition.
+// State changes should always be synchronous: only one goroutine at a time should change state.
+// n.stateMu should not be locked for long periods of time because external clients expect a timely response from n.State()
+func (n *node) setState(s NodeState) {
+	n.stateMu.Lock()
```

Comment:
There are atomic functions as well: https://pkg.go.dev/sync/atomic
But this case requires a RW lock anyways IIUC, since it doesn't reduce cleanly to CAS.

Reply:
(and we generally use [go.uber.org/atomic](https://pkg.go.dev/go.uber.org/atomic) values instead of sync/atomic)

Reply:
Ahh, thanks for the context. Another reminder I need to read the go docs in depth.

Reply:
This three part series is really informative, and this part is particularly relevant: https://research.swtch.com/gomm#maybe